### PR TITLE
chore(storage): introduce JournalIndex injection

### PR DESCRIPTION
**Pull request by ChrisKujawa**
In order to inject own JournalIndex from the outside (from Zeebe) we
 introduce a JournalIndex Supplier which is used in the Journal.
 The RaftPartition, Log etc pass the Supplier to the SegmentedJournal
 which uses that on creating new segments to pass an JournalIndex

 The JournalIndex interface got new method to compact the index. This
 means it removes stored indices lower then the given one. This is the
 opposite operation of truncate.

 On Indexing we use the complete Entry, such that it is possible for
 other Index implementations also take the complete entry into
 account.

This is the preparation for https://github.com/zeebe-io/zeebe/issues/3623

@npepinpe @deepthidevaki actually I don't care who reviews it since you both know about the current topic :smile:

### CODE REVIEW

File: storage/src/main/java/io/atomix/storage/journal/index/JournalIndex.java [hunk 1]
```diff
@@ -37,10 +39,23 @@
...
+   * Compacts the index until the next stored index (exclusively), which means everything lower then
+   * the stored index will be removed.
+   *
+   * <p>Example Index: {5 -> 10; 10 -> 20; 15 -> 30}, when compact is called with index 11.
+   * The next lower stored index is 10, everything lower then this index will be removed. This means
+   * the mapping {5 -> 10}, should be removed.
+   *
+   * @param index the index to which to compact the index
+   */
+  void compact(long index);
```
Comment (npepinpe) [inline storage/src/main/java/io/atomix/storage/journal/index/JournalIndex.java:59]: Wait, so we never cleared the index after compaction? :scream:

File: storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java [hunk 2]
```diff
@@ -86,18 +88,21 @@ public SegmentedJournal(
...
     this.directory = checkNotNull(directory, "directory cannot be null");
     this.namespace = checkNotNull(namespace, "namespace cannot be null");
     this.maxSegmentSize = maxSegmentSize;
     this.maxEntrySize = maxEntrySize;
     this.maxEntriesPerSegment = maxEntriesPerSegment;
-    this.indexDensity = indexDensity;
     this.flushOnCommit = flushOnCommit;
     journalMetrics = new JournalMetrics(name);
+    this.journalIndexFactory =
+        journalIndexFactory == null ? () -> new SparseJournalIndex(indexDensity)
```
Comment (npepinpe) [inline storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java]: Why not have the builder's `withIndexDensity` just set the factory to `() -> new SparseJournalIndex(density)`? Do we still need these two parameters?

File: storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java [hunk 3]
```diff
@@ -467,8 +475,10 @@ synchronized void removeSegment(JournalSegment segment) {
    * @param descriptor The segment descriptor.
    * @return The segment instance.
    */
-  protected JournalSegment<E> newSegment(JournalSegmentFile segmentFile, JournalSegmentDescriptor descriptor) {
-    return new JournalSegment<>(segmentFile, descriptor, storageLevel, maxEntrySize, indexDensity, namespace);
+  protected JournalSegment<E> newSegment(JournalSegmentFile segmentFile,
+      JournalSegmentDescriptor descriptor) {
+    return new JournalSegment<>(segmentFile, descriptor, storageLevel, maxEntrySize, namespace,
+        journalIndexFactory.get());
```
Comment (npepinpe) [inline storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java:480]: Does it make sense to get a new factory every time when getting a new segment? Why not just have the journal factory in the constructor and use the same journal index always?

File: storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java [hunk 4]
```diff
@@ -621,11 +635,13 @@ public long getCompactableIndex(long index) {
...
     Map.Entry<Long, JournalSegment<E>> segmentEntry = segments.floorEntry(index);
     if (segmentEntry != null) {
-      SortedMap<Long, JournalSegment<E>> compactSegments = segments.headMap(segmentEntry.getValue().index());
+      SortedMap<Long, JournalSegment<E>> compactSegments = segments
+          .headMap(segmentEntry.getValue().index());
       if (!compactSegments.isEmpty()) {
         log.debug("{} - Compacting {} segment(s)", name, compactSegments.size());
         for (JournalSegment segment : compactSegments.values()) {
           log.trace("Deleting segment: {}", segment);
+          segment.compactIndex(index);
```
Comment (npepinpe) [inline storage/src/main/java/io/atomix/storage/journal/SegmentedJournal.java:643]: Nice catch :muscle:

**Review by npepinpe — CHANGES_REQUESTED** [review]
:+1: 

Looks good. I would propose two things: 

1. Use the factory everywhere - so where we specify the index density, just have it use a `SparseJournalIndexFactory` with that density
1. Not call the supplier on every new segment, but just once when we create the journal - or is this necessary?

**Comment by ChrisKujawa** [comment]
Hey thanks for the quick review @npepinpe 

> Looks good. I would propose two things:
> 1. Use the factory everywhere - so where we specify the index density, just have it use a SparseJournalIndexFactory with that density
> 2. Not call the supplier on every new segment, but just once when we create the journal - or is this necessary?

1. I found that the density is more a property of the SparseJournalIndex unrelated to the JournalIndex, if you accept it on the factory you make it a property of the JournalIndex itself. 
Before I was not think that I need it, but in my current implementation i also use it so yes I can change it if you want.
2. I didn't wanted to change to much. The JournalIndex was created before for each Segment, which is the reason why I build it that way to have the smallest changes possible. It is also the reason why there was no compact before, since each segment has its own index.

**Comment by npepinpe** [comment]
Ah, so `JournalIndex` is really `JournalSegmentIndex`? :see_no_evil: 

I think regarding 1. I didn't explain what I mean - we both agree, it's a property of `SparseJournalIndex`. So why not have a `SparseJournalIndexFactory` which is created with the density? And set that factory in the journal, instead of passing both the index density and a null journal index factory when we want to use the sparse journal index?

**Comment by ChrisKujawa** [comment]
Yes you're right. I  have overlooked that in my tunnel yesterday. Thanks for pointing this out. I removed the density completely.

**Review by npepinpe — APPROVED** [review]
:+1:

**Comment by ChrisKujawa** [comment]
Bors r+

**Pull request merged**